File: simple.cpp
// Transcription into class style of c++
// C++11 features reduced to work with gcc 4.6
// compile with g++ -std=c++0x

#include <functional>
#include <string>
#include <tuple>
#include <memory>

using std::function;
using std::string;
using std::tuple;
using std::make_tuple;
using std::unique_ptr;

// Missed out of the standard
template<typename T, typename ...Args>
std::unique_ptr<T> make_unique( Args&& ...args )
{
  return std::unique_ptr<T>( new T( std::forward<Args>(args)... ) );
}

class Value {
public:
  virtual float asfloat() const = 0;
};

class Int : public Value {
  int value;
  float asfloat() const /*override*/ {return value;}
public:
  Int(int v) : value(v) {};
};

class Float : public Value {
  double value;
  float asfloat() const /*override*/ {return value;}
public:
  Float(float v) : value(v) {};
};

auto asfloat = [](const Value& v) {
  return v.asfloat();
};


typedef
tuple<string, int> binop_info;

class Binop {
public:
  virtual binop_info lookup_binop() const = 0;
  virtual unique_ptr<Value> eval(const Value&, const Value&) const = 0;
};

class Plus: public Binop {
  binop_info lookup_binop() const {
    return make_tuple("+", 0);
  }
  /*
  unique_ptr<Value> eval(const Value&, const Value&) const {
  }
  */
};

class Minus: public Binop {
  binop_info lookup_binop() const {
    return make_tuple("-", 0);
  }
};

class Multiply: public Binop {
  binop_info lookup_binop() const {
    return make_tuple("*", 2);
  }
};

class Divide: public Binop {
  binop_info lookup_binop() const {
    return make_tuple("/", 2);
  }
};

class Power: public Binop {
  binop_info lookup_binop() const {
    return make_tuple("**", 4);
  }
};

auto lookup_binop = [](const Binop& b) {
  return b.lookup_binop();
};

// 
enum UnOp {
  Negate, Sqr, Sqroot
};


#include <iostream>

using std::cout;
int main() {
  cout << std::showpoint;
  cout << asfloat(Int(3)) << "\n";
  cout << asfloat(Float(17.d)) << "\n";
};

File: simple.ml
type value =
    Int of int
  | Float of float

type binop =
    Plus
  | Minus
  | Multiply
  | Divide
  | Power

let asfloat v =
  match v with
      Int i -> (float i)
    | Float f -> f

let lookup_binop =
  let dobinop iop fop v1 v2 =
    match (v1, v2) with
        (Int i, Int j) -> Int (iop i j)
      | _  -> Float ( fop (asfloat v1) (asfloat v2))
  in
  function
    | Plus -> "+",0, dobinop (+) (+.)
    | Minus -> "-",0, dobinop (-) (-.)
    | Multiply -> "*",2, dobinop ( * ) ( *. )
    | Divide -> "/",2, dobinop (/) (/.)
    | Power -> "**",4, 
        (function x -> function y -> Float( (asfloat x)**(asfloat y)) )

type unop =
    Negate
  | Square
  | Sqroot

let lookup_unop =
  let dounop iop fop v = match v with
      Int i -> Int (iop i)
    | Float f -> Float (fop f)
  in
  let thingyop op v = op v v
  in
  function
    | Negate -> "~-", dounop (~-) (~-.)
    | Square -> "sqr", dounop (thingyop ( * )) (thingyop ( *. ))
    | Sqroot -> "sqroot", function v -> Float (sqrt (asfloat v))

type expression =
    BinOp of binop * expression * expression
  | UnOp of unop * expression
  | Value of value

let rec eval e =
  let evalbinop op v1 v2 =
    let _, _, dofunc = lookup_binop op
    in
    dofunc v1 v2
  in
  let evalunop op v =
    let _, dofunc = lookup_unop op
    in
    dofunc v
  in
  match e with
      BinOp (op, e1, e2) -> evalbinop op (eval e1) (eval e2)
    | UnOp (op, e) -> evalunop op (eval e)
    | Value (v) -> v

let string_of_exp e =
  let rec string_of_exp_b e p =
    let dobraces p1 p2 s =
      if p1 < p2
      then "("^s^")"
      else s
    in
    let string_of_binop op e1 e2 p =
      let s, pred, _ = lookup_binop op
      in
      dobraces pred p
        ((string_of_exp_b e1 pred)^s^(string_of_exp_b e2 pred))
    in
    let string_of_unop op e p =
      let s, _ = lookup_unop op
      in
      s^"("^(string_of_exp_b e p)^")"
    in
    let string_of_value v =
      match v with
          Int i -> string_of_int i
        | Float f -> string_of_float f
    in
    match e with
        BinOp (op, e1, e2) -> string_of_binop op e1 e2 p
      | UnOp (op, e) -> string_of_unop op e p
      | Value v -> string_of_value v
  in
  string_of_exp_b e 0
